Refactor and update SNOW integration v2 docs. (#1415)
hashicorp/terraform-website: content/source/docs/cloud/integrations/service-now/service-catalog.html.md

---
layout: "cloud"
page_title: "Service Catalog - ServiceNow Service Catalog Integration - Terraform Cloud and Terraform Enterprise"
description: |-
  Configure Service Catalogs with Request Items to allow end users to provision

>   infrastructure using ServiceNow and Terraform Enterprise.

Reviewer: ```suggestion
  infrastructure using ServiceNow and Terraform Cloud.
```

---

# Service Catalog Configuration

ServiceNow Service Catalogs offer Service Items to your end users. When using

> ServiceNow with the Terraform Enterprise integration, you will configure at

Reviewer: ```suggestion
ServiceNow with the Terraform Cloud integration, you will configure at
```

least one Service Catalog, with one or more Service Items in it. Each Service
Item will be configured with a Version Control System (VCS) repository
containing the Terraform configuration which will be used to provision that
infrastructure. End users will request Service Items from the Service Catalog,

> and Terraform Enterprise will fulfill the request by creating a new Workspace,

Reviewer: ```suggestion
and Terraform Cloud will fulfill the request by creating a new Workspace,
```

applying the configuration, and then reporting the results back to ServiceNow.

## Prerequisites

> 1. The Terraform Enterprise integration software must be installed and

Reviewer: ```suggestion
1. The Terraform Cloud integration software must be installed and
```

   configured on your ServiceNow instance. These steps are covered in the
   [installation documentation](./index.html).
1. You will need the The OAuth Token ID and repository identifier for each VCS

>    repository that Terraform Enterprise will use to provision infrastructure.

Reviewer: ```suggestion
   repository that Terraform Cloud will use to provision infrastructure.
```

>    Your Terraform Enterprise Admin will provide these to you.

Reviewer: ```suggestion
   Your Terraform Cloud Admin will provide these to you.
```

Once these steps are complete, in order for end users to provision

> infrastructure with ServiceNow and Terraform Enterprise, the ServiceNow Admin

Reviewer: ```suggestion
infrastructure with ServiceNow and Terraform Cloud, the ServiceNow Admin
```

   1. Click the plus sign in the top right.
   1. Select "Catalogs > Terraform Catalog > Title and Image" and choose a
      location to add the Service Catalog.
   1. Close the "Sections" dialog box by clicking the "x" in the upper right-hand
      corner.

-> **Note:** In step 1, be sure to choose "Catalogs", not "Catalog" from the
left-hand navigation.

## Configure VCS Repositories

-> **ServiceNow Roles:** `admin` or `x_terraform.vcs_repositories_user`

To make infrastructure available to your users, you must add one or more VCS
repositories containing Terraform configuration to the Terraform service
catalog. A single VCS repository can be used for multiple Service Items,
potentially with different variable values.

1. In ServiceNow, open the "Terraform > VCS Repositories" table by searching for
   "terraform" in the left-hand navigation.
1. Click "New" to add a VCS repository.
    - Name: The name for this repository. This name will be visible to end
      users, and does not have to be the same as the repository name as defined
      by your VCS provider. Ideally it will succinctly describe the
      infrastructure that will be provisioned by Terraform from the repository.

>     - OAuth Token ID: The OAuth Token ID that from your Terraform Enterprise

Reviewer: ```suggestion
    - OAuth Token ID: The OAuth Token ID that from your Terraform Cloud
```

      organization's VCS providers settings. This ID specifies which VCS

>       provider configured in Terraform Enterprise hosts the desired repository.

Reviewer: ```suggestion
      provider configured in Terraform Cloud hosts the desired repository.
```

    - Identifier: The VCS repository that contains the Terraform configuration
      for this workspace template. Repository identifiers are determined by your
      VCS provider; they typically use a format like
      `<ORGANIZATION>/<REPO_NAME>` or `<PROJECT KEY>/<REPO_NAME>`. Azure DevOps
      repositories use the format `<ORGANIZATION>/<PROJECT>/_git/<REPO_NAME>`.
    - The remaining fields are optional.
      - Branch: The branch within the repository, if different from the default
        branch.
      - Working Directory: The directory within the repository containing
        Terraform configuration.
      - Terraform Version: The version of Terraform to use. This will default to

>         the latest version of Terraform supported by your Terraform Enterprise

Reviewer: ```suggestion
        the latest version of Terraform supported by your Terraform Cloud
```

        instance.
1. Click "Submit".

![Screenshot: ServiceNow New VCS Repository](./images/service-now-vcs-repository.png "Screenshot of the ServiceNow Terraform New VCS Repository page")

> **FIXME:** The current instructions seem to stop before I can actually provision

Reviewer: LMK if the offline doc I provided doesn't answer this

infrastructure. Among other things, I don't see anything telling me how to
connect the VPC repository to the catalog item. I tried looking at steps from
the v1 integration page (below), but this doesn't seem quite right.

> **FIXME:** Related, the documentation on configuring variables and variable sets

Reviewer: @pwelch ?

Author: I think I've got this sorted out, but would still like some more eyes on it!

---
layout: "cloud"
page_title: "Service Catalog - ServiceNow Service Catalog Integration - Terraform Cloud and Terraform Enterprise"
description: |-
  Configure Service Catalogs and VCS repositories to allow end users to
  provision infrastructure using ServiceNow and Terraform Cloud.
---

# Service Catalog Configuration

When using ServiceNow with the Terraform Cloud integration, you will configure

> at least one Service Catalog. You will also configure one or more Version
> Control System (VCS) repositories containing the Terraform configuration which

Reviewer 2: ```suggestion
at least one Service Catalog. You will also configure one or more version
control system (VCS) repositories containing the Terraform configuration which
```

nocaps

will be used to provision that infrastructure. End users will request
infrastructure from the Service Catalog, and Terraform Cloud will fulfill the

> request by creating a new Workspace, applying the configuration, and then

Reviewer 2: nocaps

reporting the results back to ServiceNow.

> ## Prerequisites

Reviewer 2: The list items below here are inconsistent. First one is a sentence describing a requirement, next two are just names of objects.

Author: Hm. Good point. I went with the following, but it still seems a little awkward. Open to suggestions.

```
1. The Terraform Cloud integration software installed and configured on
   your ServiceNow instance. These steps are covered in the [installation
   documentation](./index.html).
```

1. The Terraform Cloud integration software must be installed and configured on
   your ServiceNow instance. These steps are covered in the [installation
   documentation](./index.html).

> 1. The The OAuth Token ID and repository identifier for each VCS repository that

Reviewer 2: ```suggestion
1. The OAuth Token ID and repository identifier for each VCS repository that
```

happens to all of us 😝

Reviewer 2: Also, nocaps on "token"

infrastructure with ServiceNow and Terraform Cloud, the ServiceNow Admin will
perform the following steps to make Service Items available to your end users.

1. Add at least one Service Catalog for use with Terraform.
1. Configure at least one one VCS repository in ServiceNow.
1. Create variable sets to define Terraform and environment variables to be used
   by Terraform Cloud to provision infrastructure.

## Add the Terraform Service Catalog

-> **ServiceNow Role:** `admin`

First, add a Service Catalog for use with the Terraform integration. Depending
on your organization's needs, you might use a single service catalog, or
several. If you already have a Service Catalog to use with Terraform, skip to
the next step.

1. In ServiceNow, open the Service Catalog > Catalogs view by searching for
   "Service Catalog" in the left-hand navigation.
   1. Click the plus sign in the top right.
   1. Select "Catalogs > Terraform Catalog > Title and Image" and choose a
      location to add the Service Catalog.
   1. Close the "Sections" dialog box by clicking the "x" in the upper right-hand
      corner.

-> **Note:** In step 1, be sure to choose "Catalogs", not "Catalog" from the

> left-hand navigation.

Reviewer 2: love it 😝

Author: ...because I did this So Many Times.

## Configure VCS Repositories

-> **ServiceNow Roles:** `admin` or `x_terraform.vcs_repositories_user`

To make infrastructure available to your users, you must add one or more VCS

> repositories containing Terraform configuration to the Terraform service

Reviewer 2: ```suggestion
repositories containing Terraform configurations to the Terraform service
```

Caught a couple of these so far, maybe should just make a general comment: the collective noun for an unknown number of terraform configs is "Terraform configurations," not "...configuration".

catalog.

1. In ServiceNow, open the "Terraform > VCS Repositories" table by searching for
   "terraform" in the left-hand navigation.

> 1. Click "New" to add a VCS repository.

Reviewer 2: ```suggestion
1. Click "New" to add a VCS repository, and fill in the following fields:
```

      organization's VCS providers settings. This ID specifies which VCS
      provider configured in Terraform Cloud hosts the desired repository.
    - Identifier: The VCS repository that contains the Terraform configuration
      for this workspace template. Repository identifiers are determined by your
      VCS provider; they typically use a format like
      `<ORGANIZATION>/<REPO_NAME>` or `<PROJECT KEY>/<REPO_NAME>`. Azure DevOps
      repositories use the format `<ORGANIZATION>/<PROJECT>/_git/<REPO_NAME>`.
    - The remaining fields are optional.
      - Branch: The branch within the repository, if different from the default
        branch.
      - Working Directory: The directory within the repository containing
        Terraform configuration.
      - Terraform Version: The version of Terraform to use. This will default to
        the latest version of Terraform supported by your Terraform Cloud
        instance.
1. Click "Submit".

![Screenshot: ServiceNow New VCS Repository](./images/service-now-vcs-repository.png "Screenshot of the ServiceNow Terraform New VCS Repository page")

## Configure a Variable Set

> Most Terraform configuration can be customized with Terraform variables or

Reviewer 2: s

environment variables. You can create a Variable Set within ServiceNow to

> contain the variables needed for a set of configuration. Your Terraform Admin

Reviewer 2: ```suggestion
contain the variables needed for a given configuration. Your Terraform Admin
```

Conversely, we use the singular noun for "the pile of .tf files that gets treated as a single workspace or working directory."

should provide these to you.

1. In ServiceNow, open the "Service Catalog > Variable Sets" table by searching for
   "variable sets" in the left-hand navigation.
1. Click "New" to add a Variable Set.
1. Select "Single-Row Variable Set".
    - Title: User-visible title for the variable set.
    - Internal name: The internal name for the variable set.
    - Order: The order in which the variable set will be displayed.
    - Type: Should be set to "Single Row"
    - Application: Should be set to "Terraform"
    - Display title: Whether the title is displayed to the end user.
    - Layout: How the variables in the set will be displayed on the screen.
    - Description: A long description of the variable set.
1. Click "Submit" to create the variable set.

> 1. Find and click on the Title of the new variable set in the Variable Sets

Reviewer 2: arguably nocaps on "title", but I'll accept whatever your best judgement is on these servicenow objects.

   table.
1. At the bottom of the variable set details page, click "New" to add a new

>    variable.

Reviewer 2: This step can be repeated several times, right? Should mention that somehow.

  - Type: Should be "Single Line Text" for most variables, or "Masked" for
    variables containing sensitive values.
  - Question: The user-visible question or label for the variable.

>   - Name: The internal name of the variable. This must match the name of the

Reviewer 2: ```suggestion
  - Name: The internal name of the variable. This must be derived from the name of the
```

Worth being more explicit that you need to mangle it, just so it doesn't seem to contradict the explanation below.

>       Terraform or environment value. Consult the table below to determine the

Reviewer 2: ```suggestion
      Terraform or environment variable. Consult the table below to determine the
```

      proper prefix for each variable name.
  - Tooltip: A tooltip to display for the variable.
  - Example Text: Example text to show in the variable's input box.

> 1. Under the "Default Value" tab, you can set a default value for the variable.

Reviewer 2: Isn't this part of the previous step? Or nah?

Author: Yeah, it's a little awkward, since the previous list is just bullets, but I think it makes sense in context? Or I could put the entire "make a variable" steps under another level that says "Make as many variables as you need to", but I don't like making super nested lists either.

When the Terraform integration applies configuration, it will map ServiceNow
variables to Terraform and environment variables using the following convention.
ServiceNow variables that begin with "sensitive_" will be saved as sensitive
variables with Terraform Cloud.

ServiceNow Variable Name | Terraform Cloud Variable
--|--
`tf_var_VARIABLE_NAME` | Terraform Variable: `VARIABLE_NAME`
`tf_env_ENV_NAME` | Environment Variable: `ENV_NAME`
`sensitive_tf_var_VARIABLE_NAME` | Sensitive Terraform Variable (Write Only): `VARIABLE_NAME`
`sensitive_tf_env_ENV_NAME` | Sensitive Environment Variable (Write Only): `ENV_NAME`

## Provision Infrastructure

Once a Service Catalog and VCS repositories have been configured, ServiceNow
users will be able to request infrastructure to be provisioned by Terraform
Cloud.

These requests will be fulfilled by Terraform Cloud, which will:

> 1. Create a new Workspace using the VCS repository provided by ServiceNow.

Reviewer 2: nocaps workspace

1. Configure variables for that Workspace, also provided by ServiceNow.

> 1. Plan and Apply the change.

Reviewer 2: nocaps apply

---
layout: "cloud"
page_title: "Service Catalog - ServiceNow Service Catalog Integration - Terraform Cloud and Terraform Enterprise"
description: |-
  Configure Service Catalogs and VCS repositories to allow end users to
  provision infrastructure using ServiceNow and Terraform Cloud.
---

# Service Catalog Configuration

When using ServiceNow with the Terraform Cloud integration, you will configure
at least one Service Catalog. You will also configure one or more version
control system (VCS) repositories containing the Terraform configurations which
will be used to provision that infrastructure. End users will request
infrastructure from the service catalog, and Terraform Cloud will fulfill the
request by creating a new workspace, applying the configuration, and then
reporting the results back to ServiceNow.

## Prerequisites

> 1. The Terraform Cloud integration software installed and configured on
>    your ServiceNow instance. These steps are covered in the [installation
>    documentation](./index.html).
> 1. The OAuth token ID and repository identifier for each VCS repository that
>    Terraform Cloud will use to provision infrastructure. Your Terraform Admin
>    will provide these to you.
> 1. Any Terraform or environment variables required by the configurations in the
>    given VCS repositories.

Reviewer 2: Yeah, they're just different kinds of items. How about smashing it apart, like this? 

```suggestion
Before continuing in this document, make sure that the Terraform Cloud integration software
is installed and configured on your ServiceNow instance. These steps are covered in the [installation
documentation](./index.html).

Additionally, you must have have the following information: 

1. The OAuth token ID and repository identifier for each VCS repository that
   Terraform Cloud will use to provision infrastructure. Your Terraform Admin
   will provide these to you.
1. Any Terraform or environment variables required by the configurations in the
   given VCS repositories.
```

variables to Terraform and environment variables using the following convention.
ServiceNow variables that begin with "sensitive_" will be saved as sensitive
variables within Terraform Cloud.

ServiceNow Variable Name | Terraform Cloud Variable
--|--
`tf_var_VARIABLE_NAME` | Terraform Variable: `VARIABLE_NAME`
`tf_env_ENV_NAME` | Environment Variable: `ENV_NAME`
`sensitive_tf_var_VARIABLE_NAME` | Sensitive Terraform Variable (Write Only): `VARIABLE_NAME`
`sensitive_tf_env_ENV_NAME` | Sensitive Environment Variable (Write Only): `ENV_NAME`

## Provision Infrastructure

Once a Service Catalog and VCS repositories have been configured, ServiceNow
users will be able to request infrastructure to be provisioned by Terraform
Cloud.

These requests will be fulfilled by Terraform Cloud, which will:

1. Create a new workspace using the VCS repository provided by ServiceNow.
1. Configure variables for that workspace, also provided by ServiceNow.
1. Plan and apply the change.
1. Report the results, including any outputs from Terraform, to ServiceNow.

Once this is complete, ServiceNow will reflect that the Request Item has been
provisioned.

> -> **Note:** Currently, the integration defaults to creating workspaces with

Reviewer: We recently got a comment that this note is confusingly worded. I wonder if we can clarify it a bit, though to me it reads fine; I may be biased since I know what it's trying to say already.

Author: Yeah, I took a shot at wording some better words:
```
-> **Note:** The integration creates workspaces with
[auto-apply](../../workspaces/settings.html#auto-apply-and-manual-apply)
enabled. Terraform Cloud will queue an apply for these workspaces whenever
changes are merged to the associated VCS repositories. This is known as the
[VCS-driven run workflow](../../run/ui.html).
```

Reviewer: Hm, this is better but it doesn't get the core point that this will affect *all* workspaces created from that repo. In the scenario where VCS-repo-1 has 10 workspaces attached, all 10 will run on commit, and they'll all have auto-apply enabled. So this is sort of a footgun warning.

Author: I added this sentence to the end of the above: "It is important to keep in mind
that all of the ServiceNow workspaces connected to a given repository will be
updated whenever changes are merged to the associated branch in that repository."
